### PR TITLE
feat: support pod injection to deploy Layotto as a sidecar in Kubernetes

### LINKED ISSUE #910: Support Pod Injection to deploy Layotto as a sidecar in Kubernetes.
**What would you like to be added**:

Support pod injection to deploy Layotto as a sidecar in Kubernetes.

**Why is this needed**:

Deploying Layotto as a sidecar is a common scenario in Kubernetes. 

We need to provide such things in Layotto:

#### Command Line Tool
A command lint tool which supports manually inject like `istioctl kube-inject`:

``` shell
lyoctl kube-inject -f kube-resources.yaml \
--injectConfigFile inj-template.tmpl \
--layottoConfigFile layotto.json \
--valuesFile values.json
```

#### MutatingWebhook 
An optional component to dynamically inject Layotto when target pod is creating.

Auto inject into pods conditions: 
+ Namespace with label: `mosn-injection: enabled` 
+ Pod without label: `sidecar.mosn.io/inject: false`

> Please refer to injection logic in Istio.

### PR DESCRIPTION
**Pull request by xiaoxiang10086**
**What this PR does**:
支持 Layotto 以 sidecar 形态注入到 Kubernetes Pod 中

**Which issue(s) this PR fixes**:

Fixes #910

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Review by Xunzhuo — COMMENTED** [review]
thanks for working on it, but plz add a user-guide docs to thie feature along with this PR, thanks.

**Comment by xiaoxiang10086** [comment]
> thanks for working on it, but plz add a user-guide docs to thie feature along with this PR, thanks.

**Comment by xiaoxiang10086** [comment]
> thanks for working on it, but plz add a user-guide docs to thie feature along with this PR, thanks.

For specific user documentation, please refer to [README.md](https://github.com/mosn/layotto/blob/adb65e4e9ff683717d6b7253b2537cbbb4b4daff/demo/state/k8s/README.md). 

In this document I demonstrate how to use Helm to deploy the Layotto sidecar injector and deploy a Layotto application to interact with Layotto on kubernetes

**Review by Xunzhuo — COMMENTED** [review]
Some points:
+ Use Docker Hub OCI to manage layotto injector helm charts and update related demos/docs
+ Add Sleep in demo to prevent shutting down container.
+ Add layotto dockerfile under docker/layotto_injector
+ Add build and push steps in layotto-ci workflows to build and push latest layotto-injector image
+ Add package and push steps in layotto-ci workflows to push latest layotto-injector helm chart in oci docker hub
+ Add release workflows for layotto-injector to release image and helm charts.

File: docker/layotto_injector/Dockerfile
Comment (Xunzhuo) [inline docker/layotto_injector/Dockerfile:7]: `make image IMAGES=layotto_injector`

Make sure this command works well. Refer to `docker/layotto/dockerfile`.

File: .github/workflows/layotto-release.yml
Comment (Xunzhuo) [inline .github/workflows/layotto-release.yml:313]: `VERSION` here should not be latest, this is a release workflow

File: .github/workflows/layotto-release.yml
Comment (Xunzhuo) [inline .github/workflows/layotto-release.yml:342]: ditto

File: .github/workflows/layotto-release.yml
Comment (Xunzhuo) [inline .github/workflows/layotto-release.yml:355]: ditto

**Review by Xunzhuo — CHANGES_REQUESTED** [review]

**Review by Xunzhuo — COMMENTED** [review]
```
❯ kgpo -w
NAME                                      READY   STATUS              RESTARTS   AGE
layotto-state-sdk-demo-6d8b87b7cd-qd6s7   0/2     ContainerCreating   0          4s
redis-master-0                            1/1     Running             0          6m42s
redis-replicas-0                          1/1     Running             0          6m42s
redis-replicas-1                          1/1     Running             0          5m42s
redis-replicas-2                          1/1     Running             0          5m17s
layotto-state-sdk-demo-6d8b87b7cd-qd6s7   1/2     Error               0          4s
layotto-state-sdk-demo-6d8b87b7cd-qd6s7   2/2     Running             1 (4s ago)   7s
```

Maybe we should set layotto as an init container.

File: pkg/injector/patcher/patch.go
Comment (Xunzhuo) [inline pkg/injector/patcher/patch.go:24]: `/spec/initContainers` ?

Comment (Xunzhuo) [reply pkg/injector/patcher/patch.go:24]: /cc @wenxuwan Should we use layotto as an init-container?

Comment (xiaoxiang10086) [reply pkg/injector/patcher/patch.go:24]: I checked the logs of these two containers. It may be because when the client demo sent the request, the Layotto container had not yet started the GRPC service. The following are the  of the two containers:

1. client demo container logs
```
runtime client initializing for: 127.0.0.1:34904
Start testing state_demo
panic: error saving state: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:34904: connect: connection refused"
goroutine 1 [running]:
main.testSave({0x570dd8?, 0x40000a2000?}, {0x574d98?, 0x40000d4500?}, {0xffffc01fbbba?, 0x1?}, {0x4bc5f7, 0x4}, {0x40000a2dd0, 0xb, ...})
        /app/client.go:138 +0x110
main.main()
        /app/client.go:63 +0x1c8
```

2. Layotto container logs
```
...
2023-10-19 12:26:03,669 [INFO] [runtime] init PhoneCallService
2023-10-19 12:26:03,669 [INFO] [runtime] init SmsService
2023-10-19 12:26:03,669 [INFO] [runtime] initRuntime stages cost: 50.84825ms
2023-10-19 12:26:03,671 [ERROR] [runtime][listTopicSubscriptions]error after callback: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 192.168.65.254:9999: connect: connection refused"
2023-10-19 12:26:03,673 [DEBUG] grpc server filter initialized success
2023-10-19 12:26:03,673 [DEBUG] [config] AddOrUpdateNetworkFilterFactories store network filter factories, name: grpc
2023-10-19 12:26:03,674 [INFO] [server] [conn handler] [add listener] add listener: 0.0.0.0:34904
```

Comment (xiaoxiang10086) [reply pkg/injector/patcher/patch.go:24]: I tested locally and use layotto as an init-container doesn't seem to work. The main container needs to wait for init-container to exit before it can be created. As a result, the main container is always in a pending state.

```
NAME                                      READY   STATUS      RESTARTS        AGE
layotto-state-sdk-demo-64b7c95966-49ktz   0/1     Init:0/1    0               8s
layotto-state-sdk-demo-64b7c95966-lm8l2   0/1     Init:0/1    0               8s
layotto-state-sdk-demo-64b7c95966-tb8kc   0/1     Init:0/1    0               8s
```

Comment (Xunzhuo) [reply pkg/injector/patcher/patch.go:24]: You are correct, so can you update your demo, I checked that you directly panic when creating client failed, can we retry it for some time, there are some package can do such things.

Comment (xiaoxiang10086) [reply pkg/injector/patcher/patch.go:24]: @Xunzhuo I checked the source code of layotto `client.NewClient()` and found that it internally calls `grpc.Dial` to create a connection. No GRPC call is sent.
```
conn, err := grpc.Dial(address, grpc.WithInsecure())
if err != nil {
	return nil, errors.Wrapf(err, "error creating connection to '%s': %v", address, err)
}
```
According to the client demo container logs, the panic is located in `cli.SaveState(ctx, store, key, value)`. So I used [retry-go](https://github.com/avast/retry-go) to retry after the request fails.
```
func testSave(ctx context.Context, cli client.Client, store string, key string, value []byte) {
	err := retry.Do(func() error {
		return cli.SaveState(ctx, store, key, value)
	},
		retry.Attempts(3),
		retry.Delay(time.Second),
	)
	if err != nil {
		panic(err)
	}
	fmt.Printf("SaveState succeeded.key:%v , value: %v \n", key, string(value))
}
```
Do you think this is reasonable?

Comment (Xunzhuo) [reply pkg/injector/patcher/patch.go:24]: Tried and it worked well, thanks.

**Comment by Xunzhuo** [comment]
I think we need some e2e tests for this feature, but this could be a follow-up.

**Comment by Xunzhuo** [comment]
/hold
/approve

Will merge after CI passed.

**Review by Xunzhuo — APPROVED** [review]

**Pull request merged**
